packages/clay-core/src/overlay-mask/OverlayMask.tsx

```diff
@@ -0,0 +1,218 @@
+/**
+ * SPDX-FileCopyrightText: © 2022 Liferay, Inc. <https://liferay.com>
+ * SPDX-License-Identifier: BSD-3-Clause
+ */
+
+import {TInternalStateOnChange, useInternalState} from '@clayui/shared';
+import React, {useEffect, useLayoutEffect, useRef, useState} from 'react';
+
+type Bounds = {
+	height: number;
+	width: number;
+	x: number;
+	y: number;
+};
+
+const M = (x: number, y: number) => `M ${x} ${y}`;
+const L = (x: number, y: number) => `L ${x} ${y}`;
+const arc = (toX: number, toY: number) => `A 0,0 0 0 0 ${toX},${toY}`;
+
+const createClipPath = (
+	bounds: Bounds,
+	containerBounds: Bounds,
+	padding: number
+) => {
+	const parentBounds = {
+		endX: containerBounds.width,
+		endY: containerBounds.height,
+		startX: 0,
+		startY: 0,
+	};
+
+	const startX = bounds.x - padding;
+	const endX = startX + bounds.width + padding * 2;
+	const startY = bounds.y - padding;
+	const endY = startY + bounds.height + padding * 2;
+
+	return [
+		M(parentBounds.startX, parentBounds.startY),
+		L(parentBounds.startX, parentBounds.endY),
+		L(parentBounds.endX, parentBounds.endY),
+		L(parentBounds.endX, parentBounds.startY),
+		'z',
+		M(startX, startY),
+		L(startX, endY),
+		arc(startX, endY),
+		L(endX, endY),
+		arc(endX, endY),
+		L(endX, startY),
+		arc(endX, startY),
+		L(startX, startY),
+		arc(startX, startY),
+	].join(' ');
+};
+
+type Props<T> = {
+	/**
+	 * Sets the current value of bounds to define the highlight area (controlled).
+	 */
+	bounds?: Bounds;
+
+	/**
+	 * Sets the element that will receive the highlight.
+	 */
+	children?: React.ReactNode | ((ref: React.RefObject<T>) => React.ReactNode);
+
+	/**
+	 * Callback is called when the overlay is clicked.
+	 */
+	onClick?: (event: React.MouseEvent<SVGRectElement>) => void;
+
+	/**
+	 * Set the highlight padding.
+	 */
+	padding?: number;
+
+	/**
+	 * Sets the current visibility of the overlay.
+	 */
+	visible?: boolean;
+
+	/**
+	 * Sets the default value of bounds (uncontrolled).
+	 */
+	defaultBounds?: Bounds;
+
+	/**
+	 * Callback is called when the bounds changes (controlled).
+	 */
+	onBoundsChange?: TInternalStateOnChange<Bounds>;
+};
+
+const initialBounds = {
+	height: 0,
+	width: 0,
+	x: 0,
+	y: 0,
+};
+
+const useIsomorphicLayoutEffect =
```

Comment:
Are we rendering this component on server side too? Sorry for the nit but I would know

Reply:
Yeah, `useLayoutEffect` doesn't render on the server causes some errors, basically it came from here #3956